Applications/Classification/jni/main_train.cpp

```diff
@@ -0,0 +1,395 @@
+ *              The base model for feature extractor is mobilenet v2 with
+ * 1280*7*7 feature size. It read the Classification.ini in res directory and
+ * run according to the configureation.
+ *
+ */
+
+#include "bitmap_helpers.h"
+#include "tensorflow/contrib/lite/interpreter.h"
+#include "tensorflow/contrib/lite/kernels/register.h"
+#include "tensorflow/contrib/lite/model.h"
+#include "tensorflow/contrib/lite/string_util.h"
+#include "tensorflow/contrib/lite/tools/gen_op_registration.h"
+#include <cmath>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <sstream>
+#include <stdlib.h>
+#include <time.h>
+
+#include "databuffer.h"
+#include "layers.h"
+#include "neuralnet.h"
+#include "tensor.h"
+
+/**
+ * @brief     Data size for each category
+ */
+#define TOTAL_TRAIN_DATA_SIZE 100
+
+#define TOTAL_VAL_DATA_SIZE 10
+
+#define TOTAL_TEST_DATA_SIZE 100
+
+#define BUFFER_SIZE 100
+
+/**
+ * @brief     Number of category : Three
+ */
+#define TOTAL_LABEL_SIZE 10
+
+/**
+ * @brief     Number of Test Set
+ */
+#define TOTAL_TEST_SIZE 8
+
+/**
+ * @brief     Max Epoch
+ */
+#define ITERATION 3000
+
+#define MINI_BATCH 32
+
+#define training true
+
+#define FEATURE_SIZE 62720
+
+using namespace std;
+
+/**
+ * @brief     location of resources ( ../../res/ )
+ */
+string data_path;
+
+bool duplicate[TOTAL_LABEL_SIZE * TOTAL_TRAIN_DATA_SIZE];
+bool valduplicate[TOTAL_LABEL_SIZE * TOTAL_VAL_DATA_SIZE];
+
+/**
+ * @brief     step function
+ * @param[in] x value to be distinguished
+ * @retval 0.0 or 1.0
+ */
+float stepFunction(float x) {
+  if (x > 0.9) {
+    return 1.0;
+  }
+
+  if (x < 0.1) {
+    return 0.0;
+  }
+
+  return x;
+}
+
+/**
+ * @brief     Get Feature vector from tensorflow lite
+ *            This creates interpreter & inference with ssd tflite
+ * @param[in] filename input file path
+ * @param[out] feature_input save output of tflite
+ */
+void getFeature(const string filename, vector<float> &feature_input) {
+  int input_size;
+  int output_size;
+  int *output_idx_list;
+  int *input_idx_list;
+  int inputDim[4];
+  int outputDim[4];
+  int input_idx_list_len = 0;
+  int output_idx_list_len = 0;
+  std::string model_path = "../../res/mobilenetv2.tflite";
+  std::unique_ptr<tflite::FlatBufferModel> model =
+    tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
+
+  assert(model != NULL);
+  tflite::ops::builtin::BuiltinOpResolver resolver;
+  std::unique_ptr<tflite::Interpreter> interpreter;
+  tflite::InterpreterBuilder(*model.get(), resolver)(&interpreter);
+
+  input_size = interpreter->inputs().size();
+  output_size = interpreter->outputs().size();
+
+  input_idx_list = new int[input_size];
+  output_idx_list = new int[output_size];
+
+  int t_size = interpreter->tensors_size();
+  for (int i = 0; i < t_size; i++) {
+    for (int j = 0; j < input_size; j++) {
+      if (strcmp(interpreter->tensor(i)->name, interpreter->GetInputName(j)) ==
```

Comment:
I think `strcmp` is unsafe. How about using `strncmp` or `std::string`'s `compare()` ?

Reply:
This going to be updated incoming PR. Thanks.